Enhancements in help message and bash completions

sshedi opened the pull request.

sshedi committed:
Enhance bash completion script

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

sshedi committed:
Enhance help message

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,6 +40,61 @@ _tdnf__process_if_prev_is_option()
+            return 0
+            ;;
+        --releasever)
+            COMPREPLY=( $(compgen -W "7 8 9" -- $cur) )
```
We only allow versions 7, 8 and 9? This is usually set to "5.0". It can be any version string. It might even be non-numeric.

oliverkurth commented:
I still see a few misalignments:
```
...
--refresh                  Refresh repository metadata before operation
                            Example: tdnf --refresh install package
...
--sec-severity <level>     Filter by security severity (Critical|Important|Moderate|Low)
                           Example: tdnf updateinfo --sec-severity=Critical
--security                  Only show security updates
                            Example: tdnf --security update
--setopt <option>=<value>  Set configuration option
                            Example: tdnf --setopt=keepcache=1 install package
--skip-broken              Skip packages with broken dependencies
                           Example: tdnf --skip-broken install pkg
```
There are more, just check.

Typo also:
```
--download-path <dir>      Directory to download packages to
                            Exampl tdnf reposync --download-path=/mnt/repo
```

I will install it also, to see if tab completion works as expected.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,6 +40,61 @@ _tdnf__process_if_prev_is_option()
             return 0
             ;;
+        --forcearch)
+            opts="x86_64 aarch64 noarch"
```
I don't think `noarch` makes sense here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,6 +40,61 @@ _tdnf__process_if_prev_is_option()
+            COMPREPLY=( $(compgen -d -- $cur) )
+            return 0
+            ;;
+        --whatdepends|--whatrequires|--whatprovides|--whatobsoletes|--whatconflicts|--whatrecommends|--whatsuggests|--whatsupplements|--whatenhances)
```
This should be packages:
```
root@photon-e440344eaaf3 [ ~ ]# tdnf repoquery --whatdepends toybox
raspberrypi-firmware-pi3-1.20230106-2.ph5.noarch
...
```

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,32 +111,99 @@ _tdnf__process_if_cmd()
-            opts=$(tdnf list --upgrades | awk '{print $1}' | cut -d'.' -f1)
+            opts=$(tdnf list --upgrades 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)
+            ;;
+        check|help|makecache|nocligpgcheck)
```
`nocligpgcheck` is an option, not a command.

oliverkurth commented on tools/cli/lib/help.c:
```diff
@@ -9,113 +9,407 @@
+ "-i, --installroot <path>   Set install root path (must be absolute)\n"
+ "                           Example: tdnf -i /mnt/chroot install pkg\n"
+
+ "--json                     Output in JSON format\n"
```
Also `-j`.

oliverkurth commented on tools/cli/lib/help.c:
```diff
@@ -9,113 +9,407 @@
+ "--nogpgcheck               Skip GPG signature checks\n"
+ "                           Example: tdnf --nogpgcheck install pkg\n"
+
+ "--nocligpgcheck            Skip GPG signature checks for local rpms\n"
```
Not local, but packages on the command line. For example, it would affect checking when run like `tdnf install --nopcligpgcheck http://foo.bar.com/package.rpm`.

Packages that are local, but in a repository (using `baseurl=file:///mylocal/rpms`) would NOT be affected.

oliverkurth commented on tools/cli/lib/help.c:
```diff
@@ -9,113 +9,407 @@
+ "--repo <repoid>            Enable only this repository\n"
+ "                           Example: tdnf --repo=base install pkg\n"
+
+ "--repofromdir <dir>        Add repository from directory\n"
```
This needs a repo id also, just like `repofrompath`.

oliverkurth commented on tools/cli/lib/help.c:
```diff
@@ -9,113 +9,407 @@
+ "                           Example: tdnf --repoid=base install pkg\n"
+
+ "--rpmdefine <macro>        Define RPM macro\n"
+ "                           Example: tdnf --rpmdefine='_prefix /usr' install package\n"
```
I think this is a bad example, I don't think it would have the expected effect. It's usually used to set the rpmdb location. I cannot think of another example, that doesn't mean they don't exist.

sshedi commented:
Thanks for the review. Fixed everything I caught, PTAL again.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,12 +39,67 @@ _tdnf__process_if_prev_is_option()
+            COMPREPLY=( $(compgen -f -- $cur) )
+            return 0
+            ;;
+        --arch)
+            opts="x86_64 aarch64"
+            COMPREPLY=( $(compgen -W "$opts" -- $cur) )
+            return 0
```
Sorry, you went too far here - for `--arch`, `noarch` is legit. It's used for `reposync` and `repoquery` to filter packages.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,12 +39,67 @@ _tdnf__process_if_prev_is_option()
+            return 0
+            ;;
+        --whatdepends|--whatrequires|--whatprovides|--whatobsoletes|--whatconflicts|--whatrecommends|--whatsuggests|--whatsupplements|--whatenhances)
+            opts=$(tdnf list --available 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)
```
Why just `--available`? It could be used for installed packages that are not available.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,12 +39,67 @@ _tdnf__process_if_prev_is_option()
+        --qf)
+            COMPREPLY=( $(compgen -W "" -- $cur) )
+            return 0
+            ;;
+        --to|--from)
+            COMPREPLY=( $(compgen -W "" -- $cur) )
+            return 0
```
This takes numbers - can we complete for those?

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +110,110 @@ _tdnf__process_if_cmd()
+        repoquery)
+            # After repoquery, offer both repoquery-specific options and package names
+            local repoquery_opts="--available --duplicates --extras --file --installed --userinstalled --upgrades --whatconflicts --whatdepends --whatenhances --whatobsoletes --whatprovides --whatrecommends --whatrequires --whatsuggests --whatsupplements --changelogs --conflicts --depends --enhances --list --location --obsoletes --provides --qf --recommends --requires --requires-pre --suggests -- ...
+            local pkg_names="$(tdnf list --available 2>/dev/null | awk '{print $1}' | cut -d'.' -f1) $(tdnf list --installed 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)"
```
I think you can just use `tdnf list` without `--available` or `--installed` and it will list all packages known. Makes this more efficient.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +110,110 @@ _tdnf__process_if_cmd()
+            ;;
+        check-update)
+            # Optional package names
+            opts="$(tdnf list --available 2>/dev/null | awk '{print $1}' | cut -d'.' -f1) $(tdnf list --installed 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)"
```
Same here, and all similar cases.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -40,12 +39,68 @@ _tdnf__process_if_prev_is_option()
+            return 0
+            ;;
+        --whatdepends|--whatrequires|--whatprovides|--whatobsoletes|--whatconflicts|--whatrecommends|--whatsuggests|--whatsupplements|--whatenhances)
+            opts=$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)
```
Since this is on the dev branch (or >= 3.6), we can use:
`tdnf repoquery --qf "%{name}" 2>/dev/null`

Also in lots of other places. It's more efficient.

However, it means we cannot backport this to 3.5 - which may be okay, since we can update tdnf to 3.6 in Photon 5.0rev2 soon.

sshedi replied:
Yes, I understand and agree.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
             ;;
         downgrade)
-            opts=$(tdnf list --downgrades | awk '{print $1}' | cut -d'.' -f1)
+            opts=$(tdnf list --downgrades 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)
```
Use `repoquery` here.

sshedi replied:
This was missing, I added this support to repoquery.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
             ;;
         autoerase|autoremove|erase|reinstall|remove)
-            opts=$(tdnf list --installed | awk '{print $1}' | cut -d'.' -f1)
+            opts=$(tdnf list --installed 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+            return 0
+            ;;
+        distro-sync)
+            opts=$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
             ;;
         install)
-            opts="$(tdnf list --available | awk '{print $1}' | cut -d'.' -f1) $(compgen -d -G '*.rpm')"
+            opts=$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
             ;;
         update|upgrade)
-            opts=$(tdnf list --upgrades | awk '{print $1}' | cut -d'.' -f1)
+            opts=$(tdnf list --upgrades 2>/dev/null | awk '{print $1}' | cut -d'.' -f1)
```
Use `repoquery` here.

sshedi replied:
This command had an issue, fixed it.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+            ;;
+        check-update)
+            # Optional package names
+            opts="$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)"
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+        list)
+            # Scope options or package names
+            local scope_opts="installed available updates downgrades recent all"
+            local pkg_names="$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)"
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+        info)
+            # Package names or scope options
+            local scope_opts="installed available updates downgrades recent all"
+            local pkg_names="$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)"
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+            ;;
+        update-to|upgrade-to)
+            # Package names (with optional version)
+            opts="$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)"
```
Use `repoquery` here.

oliverkurth commented on etc/bash_completion.d/tdnf-completion.bash:
```diff
@@ -56,44 +111,110 @@ _tdnf__process_if_cmd()
+        updateinfo)
+            # Mode options or package names
+            local mode_opts="all info summary"
+            local pkg_names="$(tdnf list 2>/dev/null | awk '{print $1}' | sort -u | cut -d'.' -f1)"
```
Use `repoquery` here.

oliverkurth reviewed (commented):
Looks good (thank you!), but I have one more suggestion: use `repoquery --qf` instead of `awk` and `cut` for `list` commands. I didn't suggest this earlier because it requires tdnf >= 3.6, but I realized that that is actually fine.

Also, I realized that we should have a simpler output version for `repolist`, and a way to get the current history count number. We can add that in another change.

sshedi committed:
Fix repoquery --upgrades

Currently `tdnf repoquery --upgrades` shows the entire list of packages
instead of just showing uavailable updates, this patch fixes it.

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

sshedi committed:
Add repoquery downgrades support

Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>

oliverkurth commented:
Thank you, looks good!
